pkg/controller/machinepool/machinepool_controller.go

```diff
@@ -563,10 +609,18 @@ func (r *ReconcileMachinePool) syncMachineSets(
 	for i, ms := range generatedMachineSets {
 		found := false
 		for _, rMS := range remoteMachineSets.Items {
-			if matchAndMutate(pool, cd, ms, rMS) {
-				found = true
+			var err error
+			found, err = matchMachineSets(ms, rMS, logger)
+			if err != nil {
+				logger.WithError(err).Errorf("unable to match MachineSets %s and %s", ms.Name, rMS.Name)
+				continue // FIXME is this the correct behavior?
+			}
+			if found {
+				logger.Infof("Found matching remote MachineSet %s for generated MachineSet %s", rMS.Name, ms.Name)
 				objectModified := false
 				objectMetaModified := false
+				// Rename generated MachineSet if there is a name mismatch
+				ms.Name = rMS.Name
```

Comment:
Is it possible for `rMS.Name` to end up colliding with a _different_ generated mset that we'll see in another iteration, either before or after this one? I.e. renaming will result in two msets in our list with the same name?

I'm not gonna try to think through this yet, just noting it down for now.